doc/neps/roadmap.rst

```diff
@@ -0,0 +1,114 @@
+-----------------
+`NEP 22`_    `np.asduckarray()`
+
+
+Mixins like `NDArrayOperatorsMixin`:
+
+- for mutable arrays
+- for reduction methods implemented as ufuncs
+
+Better dtypes
+-------------
+
+- Easier custom dtypes
+  - Simplify and/or wrap the current C-API
+  - More consistent support for dtype metadata
+  - Support for writing a dtype in Python
+- New string dtype(s):
+  - Fixed width encoded strings (utf8, latin1, ...)
+  - Variable length strings (could share implementation with dtype=object, but are explicitly type-checked)
+  - One of these should probably be the default for text data. The current behavior on Python 3 is neither efficient nor user friendly.
+- `np.int` should not be platform dependent
+- better coercion for string + number
+
+Random number generation policy & rewrite
+-----------------------------------------
+
+`NEP 19`_ and a `reference implementation`_
+
+Indexing
+--------
+
+vindex/oindex `NEP 21`_
+
+Infrastructure
+--------------
+
+NumPy is much more than just the code base itself, we also maintain
+docs, CI, benchmarks, etc.
+
+- Rewrite numpy.org
+- Benchmarking: improve the extent of the existing suite, and run & render
+  the results as part of the docs or website.
+
+  - Hardware: find a machine that can reliably run serial benchmarks
+  - ASV produces graphs, could we set up a site? Currently at
+    https://pv.github.io/numpy-bench/, should that become a community resource?
+
+Functionality outside core
+--------------------------
+
+Some things inside NumPy do not actually match the `Scope of NumPy`.
+
+- A backend system for `numpy.fft` (so that e.g. `fft-mkl` doesn't need to monkeypatch numpy)
+
+- Rewrite masked arrays to not be a ndarray subclass -- maybe in a separate project?
+- MaskedArray as a duck-array type, and/or
+- dtypes that support missing values
+
+- Write a strategy on how to deal with overlap between numpy and scipy for `linalg` and `fft` (and implement it).
+
+- Deprecate `np.matrix`
+
+Continuous Integration
+----------------------
+
+We depend on CI to discover problems as we continue to develop NumPy before the
+code reaches downstream users.
+
+- CI for more exotic platforms (e.g. ARM is now available from
+  http://www.shippable.com/, but it is not free).
+- Multi-package testing
+- Add an official channel for numpy dev builds for CI usage by other projects so
+  they may confirm new builds do not break their package.
+
+Typing
+------
+
+Python type annotation syntax should support ndarrays and dtypes.
+
+- Type annotations for NumPy: github.com/numpy/numpy-stubs
+- Support for typing shape and dtype in multi-dimensional arrays in Python more generally
+
+NumPy scalars
+-------------
+
+Numpy has both scalars and zero-dimensional arrays.
+
+- The current implementation adds a large maintenance burden -- can we remove
+  scalars and/or simplify it internally?
+- Zero dimensional arrays get converted into scalars by most NumPy functions.
+  Can this be fixed?
```

Comment:
And vice versa, scalars are converted to 0-D arrays for determining casting and also when they are ufunc arguments. It would be good to look at making 0-D arrays more efficient.

Reply:
How do you recommend I update the text?

Reply:
Maybe "Most NumPy functions convert scalar inputs to zero dimensional arrays and return zero dimensional arrays as scalars. That back and forth is not efficient."

The scalar problem is difficult to solve for many reasons, almost all having to do with backwards compatibility and the promotion rules concerning array scalar interaction. The latter alone is a major problem and I think requires that we retain scalar types in some form. Perhaps scalars should subclass ndarrays or vice versa. I believe @teoliphant once made a comment along those lines.

Reply:
I'd argue the problem is not so much efficiency, but the inconsistency. It's very hard to correctly promote a scalar back to a 0d array of the right dtype once you've let it decay.

Reply:
I agree that inconsistency is the real problem. An illustration of this is that it's impossible to use Python typing to annotate most NumPy functions, e.g., `np.sin(array)` returns a numpy array when the input is a numpy array, unless it happens to be 0-dimensional.